Make Unet configurable

neptunes5thmoon opened the pull request:
Hi @DhairyaLGandhi 

I often need flexibility with the hyperparameters of the UNet, i.e. different numbers of downsamplings and convolutions, kernel sizes etc. So we started implementing this more configurable version. 

This currently allows for a UNet with configurable
- number of input and output channels
- initial number of feature maps
- factor by which number of feature maps should be multiplied with each downsampling
- downsampling method, factors and number of downsampling steps
- kernel sizes and number of convolutions
- activation functions
- "same" or "valid" padding

Let me know if this is something you would be interested in including here.

RainerHeintzmann committed:
Added version compatibility

neptunes5thmoon committed:
WIP configurable UNet

neptunes5thmoon committed:
add example with configured UNet

RainerHeintzmann committed:
bug fixes

RainerHeintzmann committed:
bug fixes

RainerHeintzmann committed:
added pretty print and removed debug prints.

RainerHeintzmann committed:
bug fixes in show(). removed flux_test.jl

RainerHeintzmann committed:
bug fixes in noise2noise and deconvolve

RainerHeintzmann committed:
reduced iterations in noise2noise example

neptunes5thmoon committed:
clarify UNet pretty print

neptunes5thmoon committed:
emphasize U-structure in pretty print

neptunes5thmoon committed:
feat: add implementation for valid padding

neptunes5thmoon committed:
Merge branch 'master' of github.com:DhairyaLGandhi/UNet.jl

neptunes5thmoon committed:
keep examples dependencies separate

neptunes5thmoon committed:
add Flux back

DhairyaLGandhi commented:
This sounds like. a good idea to support! I will take a look at the changes in the PR as well!

DhairyaLGandhi commented on src/model.jl:
```diff
@@ -1,106 +1,275 @@
-    Chain(Conv(kernel, in_chs=>out_chs,pad = (1, 1);init=_random_normal),
-	BatchNormWrap(out_chs),
-	x->leakyrelu.(x,0.2f0))
+struct ConvBlock
```
This should be parameterised

```suggestion
struct ConvBlock{T}
```

RainerHeintzmann replied:
done

DhairyaLGandhi commented on src/model.jl:
```diff
@@ -1,106 +1,275 @@
 
-struct UNetUpBlock
-  upsample
+function ConvBlock(in_channels, out_channels, kernel_sizes = [(3,3), (3,3)];
```
We can mirror the Flux `Conv` API here - this would mean we switch out the positions of the kernels, and make the channels as a `Pair`. This does lose out on having default kernel pairs though. Are we expecting to not need to define the kernels much? I suppose we wouldn't need to, but maintaining that consistency in the API would be a plus. What do you think?

RainerHeintzmann replied:
Yes, good idea. There is a way to allow optional first arguments, since the number of arguments changes. Therefore we made the first argument now optional, followed by the mentioned `Pair`.

DhairyaLGandhi commented on src/model.jl:
```diff
@@ -1,106 +1,275 @@
-  x = u.upsample(x)
-  return cat(x, bridge, dims = 3)
+function Base.show(io::IO, d::Downsample)
+  print(io, "Unet.Downsample($(d.factor), $(d.pooling_type))")
```
```suggestion
  print(io, "Downsample($(d.factor), $(d.pooling_type))")
```

DhairyaLGandhi commented on src/model.jl:
```diff
@@ -1,106 +1,275 @@
-			stride=(2, 2);init=_random_normal),
-		BatchNormWrap(out_chs),
-		Dropout(p)))
+struct Downsample
```
Same comment about paramterization as earlier

RainerHeintzmann replied:
done

DhairyaLGandhi commented on src/model.jl:
```diff
@@ -1,106 +1,275 @@
-  conv_down_blocks
-  conv_blocks
-  up_blocks
+function Downsample(downsample_factor; pooling_type="max")
```
If we took the function `maxpool` or `meanpool` directly, we can get rid of the conditional.

RainerHeintzmann replied:
Good idea. Changed accordingly.

DhairyaLGandhi commented on src/model.jl:
````diff
@@ -1,106 +1,275 @@
+```jldoctest
+```
+"""
+function Unet(;  # all arguments are named and ahve defaults
````
What do you think about retaining some of the positional argument versions as well?

neptunes5thmoon replied:
Do you mean as above to mirror the flux API? Or are there specific arguments you think should be positional?

DhairyaLGandhi replied:
I feel the channels can be made a Pair `in => out`

RainerHeintzmann replied:
done

neptunes5thmoon committed:
add parametrization

neptunes5thmoon committed:
declutter print

Co-authored-by: Dhairya Gandhi <dhairya@juliacomputing.com>

neptunes5thmoon committed:
only conv chains should share type

neptunes5thmoon committed:
separate conv chain types

Even for conv chains the type can't be shared since different numbers of convolutions are permissible

mkitti commented:
Following. Where did we end up here?

neptunes5thmoon commented:
Hi @mkitti 
I was waiting for some feedback to finish my updates, then got sidetracked by other projects. Feel free to add any edits/ideas/suggestions you might have!

DhairyaLGandhi commented on src/model.jl:
```diff
@@ -1,106 +1,275 @@
+  activation = NNlib.relu,
+  final_activation = NNlib.relu,
+  padding ="same",
+  pooling_type ="max"
```
I would prefer to take in the function directly if possible - that way we simply call the input function and users can specify their own polling if preferable

DhairyaLGandhi reviewed (commented):
Can confirm,  got side tracked :smile: 

I've added a couple comments, but it looks mostly good! I would love to add this in!

RainerHeintzmann committed:
Merge branch 'DhairyaLGandhi:master' into master

RainerHeintzmann committed:
updated version numbers of packages. Adapted noise2noise exampel

RainerHeintzmann committed:
updated deconvoltion.jl

RainerHeintzmann committed:
bug fixes and chages according to comments

neptunes5thmoon commented:
I think we addressed all your suggestions. Could you have another look @DhairyaLGandhi
